### PR TITLE
[Perf] Replace some exhaustive O(n^3) or O(n^2) iterations with O(n)

### PR DESCRIPTION
**Pull request by Therzok**
Usage of g_slist_count and g_slist_nth are actually O(n) operations by themselves. Nesting them can prove to be CPU intensive.

### CODE REVIEW

File: mono/metadata/w32process-unix.c [hunk 1]
```diff
@@ -770,18 +767,26 @@ mono_w32process_try_get_modules (gpointer process, gpointer *modules, guint32 si
...
+		mods_iter = g_slist_next (mods_iter);
+		count++;
 	}
 
-	for (i = 0; i < count; i++) {
-		mono_w32process_module_free ((MonoW32ProcessModule *)g_slist_nth_data (mods, i));
+	/* count + 1 to leave slot 0 for the main module */
+	*needed = sizeof(gpointer) * (count + 1);
+
+	for (mods_iter = mods; mods_iter; mods_iter = g_slist_next (mods_iter)) {
```
Comment (Therzok) [inline mono/metadata/w32process-unix.c]: Maybe this could be freed in-place?

**Comment by Therzok** [comment]
cc @kumpera ?

File: mono/metadata/w32process-unix.c (hunk 1 above)
Comment (BrzVlad) [inline mono/metadata/w32process-unix.c]: Looks like we could

**Comment by BrzVlad** [comment]
LGTM

File: mono/metadata/w32process-unix.c (hunk 1 above)
Comment (Therzok) [inline mono/metadata/w32process-unix.c]: Fixed.

**Comment by Therzok** [comment]
@BrzVlad good to merge?

**Pull request merged**